core/arch/interp.c

```diff
@@ -817,6 +817,26 @@ check_new_page_jmp(dcontext_t *dcontext, build_bb_t *bb, app_pc new_pc)
 }
 
 static inline void
+bb_process_single_step(dcontext_t *dcontext, build_bb_t *bb)
+{
+    LOG(THREAD, LOG_INTERP, 2, "interp: single step exception bb at "PFX"\n", bb->instr_start);
+    instrlist_append(bb->ilist, bb->instr);
+    /* Sets exit reason dynamically. */
+    instrlist_meta_preinsert(bb->ilist, bb->instr,
+                             instr_create_save_immed16_to_dcontext(dcontext,
+                                                                   EXIT_REASON_SINGLE_STEP,
+                                                                   EXIT_REASON_OFFSET));
```

Comment:
OK.
I tried to do this but I am not sure about when to reset `single_step_addr`